x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/CCRIT.java

```diff
@@ -0,0 +1,292 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.upgrades;
+
+import org.apache.http.util.EntityUtils;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.Version;
+import org.elasticsearch.client.Request;
+import org.elasticsearch.client.Response;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.xcontent.ObjectPath;
+import org.elasticsearch.common.xcontent.XContentHelper;
+import org.elasticsearch.common.xcontent.json.JsonXContent;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+
+import static org.hamcrest.Matchers.equalTo;
+
+public class CCRIT extends AbstractUpgradeTestCase {
+
+    private static final Logger LOGGER = LogManager.getLogger(CCRIT.class);
+
+    private static final Version UPGRADE_FROM_VERSION =
+        Version.fromString(System.getProperty("tests.upgrade_from_version"));
+
+    private static final boolean SECOND_ROUND = "false".equals(System.getProperty("tests.first_round"));
+
+    @Override
+    protected boolean preserveClusterSettings() {
+        return true;
+    }
+
+    public void testIndexFollowing() throws Exception {
+        assumeTrue("CCR became available in 6.5 and test relies on a fix that was shipped with 6.5.4",
+            UPGRADE_FROM_VERSION.onOrAfter(Version.V_6_5_4));
+        setupRemoteCluster();
+
+        final String leaderIndex = "my-leader-index";
+        final String followerIndex = "my-follower-index";
+
+        switch (CLUSTER_TYPE) {
+            case OLD:
+                Settings indexSettings = Settings.builder()
+                    .put("index.soft_deletes.enabled", true)
+                    .put("index.number_of_shards", 1)
+                    .build();
+                createIndex(leaderIndex, indexSettings);
+                followIndex(leaderIndex, followerIndex);
+                index(leaderIndex, "1");
+                assertDocumentExists(leaderIndex, "1");
+                assertBusy(() -> {
+                    assertFollowerGlobalCheckpoint(followerIndex, 0);
+                    assertDocumentExists(followerIndex, "1");
+                });
+                break;
+            case MIXED:
+                if (SECOND_ROUND == false) {
+                    index(leaderIndex, "2");
+                    assertDocumentExists(leaderIndex, "2");
```

Comment:
Maybe assert the existence of both doc1 and doc2? You can extend `assertDocumentExists` to accept varargs.